### ISSUE TITLE
Issues with running Sweeps

### ISSUE THREAD

**Issue by wiseyoungbuck** [ThilinaRajapakse/simpletransformers#545]
I am running the example Sweep here
https://github.com/ThilinaRajapakse/simpletransformers/blob/master/examples/hyperparameter%20tuning/sweeps.pyjjj

The sweep runs and all the hyperparameters are logged, however the metrics, such as training_loss are not logged. There are no summary metrics in any of the runs. All the metrics and summary metrics appear as null. 
Should simpletransformers automatically log metrics? If not how would I do that?

<img width="1275" alt="Screen Shot 2020-07-04 at 8 23 56 PM" src="https://user-images.githubusercontent.com/14169315/86523588-4ac19980-be34-11ea-9720-4f05ab8230b7.png">
<img width="1556" alt="Screen Shot 2020-07-04 at 8 24 39 PM" src="https://user-images.githubusercontent.com/14169315/86523600-662ca480-be34-11ea-96d7-0ff2e36df58d.png">

**Desktop (please complete the following information):**
 - Ubuntu

**Comment by ThilinaRajapakse** [comment]
There was a mistake in the example code (missing line) which has now been fixed. The problem is that the `wandb_project` must be specified for simple transformers to automatically log metrics and this line was missing in the example.

**Comment by aurik3l** [comment]
I still seem to get no data on the train_loss (same with eval_loss). Just to be sure, is the minimal sweeps example fully functional or is there something else that I need to do (code-wise)?

**Comment by aurik3l** [comment]
Below is the minimal example which, unfortunately, does not seem to work (slightly adapted as the calculation of MCC failed in the other example). Am I missing something here?

`import logging

import pandas as pd
import sklearn

import wandb
from simpletransformers.classification import (
    ClassificationArgs,
    ClassificationModel,
)

wandb.login();

sweep_config = {
    "method": "bayes",  # grid, random
    "metric": {"name": "train_loss", "goal": "minimize"},
    "parameters": {"num_train_epochs": {"values": [2, 3, 5]}, "learning_rate": {"min": 5e-5, "max": 4e-4}},
}

sweep_id = wandb.sweep(sweep_config, project="Simple Sweep")

# Preparing train data
train_data = [['Example sentence belonging to class 1', 1], ['Example sentence belonging to class 0', 0]]
train_df = pd.DataFrame(train_data)
train_df.columns = ["text", "labels"]

# Preparing eval data
eval_data = [['Example eval sentence belonging to class 1', 1], ['Example eval sentence belonging to class 0', 0]]
eval_df = pd.DataFrame(eval_data)
eval_df.columns = ["text", "labels"]

model_args = ClassificationArgs()
model_args.fp16 = False
model_args.reprocess_input_data = True
model_args.overwrite_output_dir = True
model_args.evaluate_during_training = True
model_args.manual_seed = 4
model_args.use_multiprocessing = True
model_args.train_batch_size = 16
model_args.eval_batch_size = 8
model_args.wandb_project = "Simple Sweep"


def train():
    # Initialize a new wandb run
    wandb.init()

    # Create a TransformerModel
    model = ClassificationModel("roberta", "roberta-base", use_cuda=True, args=model_args, sweep_config=wandb.config)

    # Train the model
    model.train_model(train_df, eval_df=eval_df)

    # Evaluate the model
    model.eval_model(eval_df)

    # Sync wandb
    wandb.join()

logging.basicConfig(level=logging.INFO)
transformers_logger = logging.getLogger("transformers")
transformers_logger.setLevel(logging.WARNING)

wandb.agent(sweep_id, train)`

**Comment by aurik3l** [comment]
Below is the minimal example that I use, which, however, does not seem to work. Do you have any ideas what might be wrong? Thank you very much!

```
import logging

import pandas as pd
import sklearn

import wandb
from simpletransformers.classification import (
    ClassificationArgs,
    ClassificationModel,
)

wandb.login();

sweep_config = {
    "method": "bayes",  # grid, random
    "metric": {"name": "train_loss", "goal": "minimize"}, # eval_loss
    "parameters": {"num_train_epochs": {"values": [2, 3, 5]}, "learning_rate": {"min": 5e-5, "max": 4e-4}},
}

sweep_id = wandb.sweep(sweep_config, project="Simple Sweep")

# Preparing train data
train_data = [['Example sentence belonging to class 1', 1], ['Example sentence belonging to class 0', 0]]
train_df = pd.DataFrame(train_data)
train_df.columns = ["text", "labels"]

# Preparing eval data
eval_data = [['Example eval sentence belonging to class 1', 1], ['Example eval sentence belonging to class 0', 0]]
eval_df = pd.DataFrame(eval_data)
eval_df.columns = ["text", "labels"]

model_args = ClassificationArgs()
model_args.fp16 = False
model_args.reprocess_input_data = True
model_args.overwrite_output_dir = True
model_args.evaluate_during_training = True
model_args.manual_seed = 4
model_args.use_multiprocessing = True
model_args.train_batch_size = 16
model_args.eval_batch_size = 8
model_args.wandb_project = "Simple Sweep"


def train():
    # Initialize a new wandb run
    wandb.init()

    # Create a TransformerModel
    model = ClassificationModel("roberta", "roberta-base", use_cuda=True, args=model_args, sweep_config=wandb.config)

    # Train the model
    model.train_model(train_df, eval_df=eval_df)

    # Evaluate the model
    model.eval_model(eval_df)

    # Sync wandb
    wandb.join()

logging.basicConfig(level=logging.INFO)
transformers_logger = logging.getLogger("transformers")
transformers_logger.setLevel(logging.WARNING)

wandb.agent(sweep_id, train)
```

**Comment by ThilinaRajapakse** [comment]
It's possibly because the minimal example only has 2 samples in the training data set. It should work with a real dataset.

You can find a full example [here](https://towardsdatascience.com/hyperparameter-optimization-for-optimum-transformer-models-b95a32b70949?source=friends_link&sk=7d19ce15c9ac1230642d826b9deeb638).

**Comment by khituras** [comment]
@aurik3l could you resolve the issue? Because I have the exact same.

I also copied the exact code for vanilla_sweep.py from

> You can find a full example [here](https://towardsdatascience.com/hyperparameter-optimization-for-optimum-transformer-models-b95a32b70949?source=friends_link&sk=7d19ce15c9ac1230642d826b9deeb638).

30 minutes ago so it should be up-to-date.

Still, while the runs seem to be logged correctly, the metrics are not logged for the sweep. The hyper parameter settings are there but the accurracy value is always 0. I would appreciate some help here a lot, I want to use this feature urgently.

**Comment by ThilinaRajapakse** [comment]
I just tested it again and it's still working ([link](https://app.wandb.ai/thilina/RTE%20-%20Hyperparameter%20Optimization%20-%20Test/sweeps/ug6w35lt?workspace=user-thilina)).

I also added the full code to the examples directory if that helps.

**Comment by khituras** [comment]
Thank you a lot for trying again.
I am really puzzled. I copied the new code from the examples directory. I had to adapt the eval_df and test_df file names because I had appended the `jsonl` extension when I did the medium article, set the batch size to 8 because I don't have enough VRAM otherwise and used the `roberta-base` model just for a quicker test.

But nothing changed, no metrics are getting logged for the training runs. Adding calls like `wandb.log({"testparam":.6})´ don't have any effect after the call to `model.train()`. However, calls done BEFORE the training are getting logged. So something in the `train()` method seems to throw the logging off course for me.
When I start a run manually and then use `wandb.log()` I get the logged metric value in the respective run in WandB, so this seems to work.

I am using version 0.45.0 of simpletransformers and 0.9.3 of wandb. Right now I have no idea what is going wrong.

**Comment by khituras** [comment]
I think I have a solution to the issue. However, I don't know why it doesn't work without that solution or I don't know why it works for @ThilinaRajapakse at all :-)

The simple solution is to pass the ID of the run created via `wandb.init()` in the user code to the newly created run in the `train` method at [line 440 in classification_model.py](https://github.com/ThilinaRajapakse/simpletransformers/blob/8750ea4cd8922b9d42b652876a22cfe2c5af3366/simpletransformers/classification/classification_model.py#L440).

This is done with:

```python
def train():
    wandb.init()
    model_args.wandb_kwargs = {"id": wandb.run.id}
    ...
```

This passes `wandb.run.id` to as a `kwarg` to the inner call to `wandb.init(...)` in `classification_model`. Now everything works fine for me.

**Comment by aurik3l** [comment]
@khituras Thank you! The code solved the issue for me.

I believe, for me, it is related to the wandb.init (line 440 of classification_model) which seems to create an additional run (without being associated with the sweep - with your code, however, it works correctly):

```
Create sweep with ID: ...
Sweep URL: https://app.wandb.ai/xyz/RTE%20-%20Hyperparameter%20Optimization/sweeps/8b2ne6on
wandb: Agent Starting Run: ... with config:
	learning_rate: 0.00026092810728350685
	num_train_epochs: 2
wandb: Agent Started Run: ...
Logging results to Weights & Biases (Documentation).
Project page: https://app.wandb.ai/xyz/RTE%20-%20Hyperparameter%20Optimization
Sweep page: https://app.wandb.ai/xyz/RTE%20-%20Hyperparameter%20Optimization/sweeps/8b2ne6on
Run page: https://app.wandb.ai/xyz/RTE%20-%20Hyperparameter%20Optimization/runs/2b8md5dx # <--- this is the correct one
Some weights of the model checkpoint at roberta-base were not used when [...]

Epoch 1 of 2: 0%
0/2 [00:03<?, ?it/s]
Logging results to Weights & Biases (Documentation).
Project page: https://app.wandb.ai/xyz/RTE%20-%20Hyperparameter%20Optimization
Run page: https://app.wandb.ai/xyz/RTE%20-%20Hyperparameter%20Optimization/runs/3jcgyk4k # <--- this one here is a new one and is not assoicated with the sweep
Epochs 0/2. Running Loss: 0.6176: 6%
3/50 [00:02<00:34, 1.37it/s]
```

**Comment by khituras** [comment]
Yes, this is what I figured and so I looked for a way to associate the runs with one another. Glad that helped!

**Comment by wiseyoungbuck** [comment]
This issues for me was caused by a hyperparameter set that was incompatible with the library and the accuracy metric was returning as NULL.

**Comment by ThilinaRajapakse** [comment]
I'm not sure why we are getting different behaviour. I asked about `wandb.init()` possibly being called more than once from people at W&B and we agreed it _probably_ won't be an issue. But, it was just a guess so we might have missed something.

Going by @khituras solution, I could just update the docs to note that the run id should be included in the `model_args` though.

**Comment by davidlenz** [comment]
here's my two cents. I just could not get a successful sweep run on my windows machine, the sweep would start until the line `... view run [url]` and then stop; the acutal runs never did anything. Tried a bunch of different things, also `wandb-ng`, nothing worked out for me. Then set up a ubuntu cloud machine and the sweeps instantly worked.

**Comment by ThilinaRajapakse** [comment]
Windows. 🤷‍♂️

**Comment by wiseyoungbuck** [comment]
After more investigate, I am no longer having this problem. I'm pretty sure it was due to runs not completing due to a bug in my code and returning null metrics.
